### PR TITLE
Implement pagination using chunks

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/storage/stream.py [hunk 1]
```diff
@@ -515,18 +538,25 @@ def _set_before_and_after(events, rows, topo_order=True):
...
-            internal.before = str(RoomStreamToken(topo, stream - 1))
-            internal.after = str(RoomStreamToken(topo, stream))
-            internal.order = (
-                int(topo) if topo else 0,
-                int(stream),
-            )
+            if topo_order and chunk:
+                internal.before = str(RoomStreamToken(chunk, topo, stream - 1))
+                internal.after = str(RoomStreamToken(chunk, topo, stream))
+                internal.order = (
```
Comment (richvdh) [inline synapse/storage/stream.py]: is wtf `order` means documented anywhere?

File: synapse/storage/stream.py [hunk 2]
```diff
@@ -515,18 +538,25 @@ def _set_before_and_after(events, rows, topo_order=True):
...
-                topo = None
+
             internal = event.internal_metadata
-            internal.before = str(RoomStreamToken(topo, stream - 1))
-            internal.after = str(RoomStreamToken(topo, stream))
-            internal.order = (
-                int(topo) if topo else 0,
-                int(stream),
-            )
+            if topo_order and chunk:
```
Comment (richvdh) [inline synapse/storage/stream.py]: is it really correct that we drop the topo ordering if we don't have a `chunk` yet for this event?

File: synapse/storage/stream.py [hunk 3]
```diff
@@ -586,27 +616,29 @@ def _get_events_around_txn(self, txn, room_id, event_id, before_limit, after_lim
...
+            retcols=["stream_ordering", "topological_ordering", "chunk_id"],
         )
 
         # Paginating backwards includes the event at the token, but paginating
         # forward doesn't.
         before_token = RoomStreamToken(
-            results["topological_ordering"] - 1,
-            results["stream_ordering"],
+            results["chunk_id"],
+            results["topological_ordering"],
```
Comment (richvdh) [inline synapse/storage/stream.py]: what's with the change to the -1 here?

File: synapse/storage/stream.py [hunk 4]
```diff
@@ -740,9 +779,68 @@ def _paginate_room_events_txn(self, txn, room_id, from_token, to_token=None,
 
         txn.execute(sql, args)
 
-        rows = [_EventDictReturn(row[0], row[1], row[2]) for row in txn]
+        rows = [_EventDictReturn(row[0], row[1], row[2], row[3]) for row in txn]
```
Comment (richvdh) [inline synapse/storage/stream.py]: isn't this normally spelt `_EventDictReturn(*row)` ?

File: synapse/storage/stream.py [hunk 5]
```diff
@@ -740,9 +779,68 @@ def _paginate_room_events_txn(self, txn, room_id, from_token, to_token=None,
...
+                chunk_id = table.get_next(chunk_id)
+
+            if chunk_id is None:
+                break
+
+            txn.execute(sql, args)
+            new_rows = [_EventDictReturn(row[0], row[1], row[2], row[3]) for row in txn]
+
+            if not new_rows:
+                break
```
Comment (richvdh) [inline synapse/storage/stream.py]: why do we do this?

File: synapse/storage/stream.py [hunk 6]
```diff
@@ -740,9 +779,68 @@ def _paginate_room_events_txn(self, txn, room_id, from_token, to_token=None,
...
+
+            if direction == 'b':
+                chunk_id = table.get_prev(chunk_id)
+            else:
+                chunk_id = table.get_next(chunk_id)
+
+            if chunk_id is None:
+                break
+
+            txn.execute(sql, args)
```
Comment (richvdh) [inline synapse/storage/stream.py]: do we not need to update the limit on each loop?

File: synapse/storage/stream.py [hunk 7]
```diff
@@ -740,9 +779,68 @@ def _paginate_room_events_txn(self, txn, room_id, from_token, to_token=None,
...
+            new_rows = [_EventDictReturn(row[0], row[1], row[2], row[3]) for row in txn]
+
+            if not new_rows:
+                break
+
+            rows.extend(new_rows)
+
+        if limit > 0:
+            rows = rows[:limit]
```
Comment (richvdh) [inline synapse/storage/stream.py]: why is this necessary (and, if it is necessary, why is it conditional on limit > 0) ?

File: synapse/storage/stream.py [hunk 8]
```diff
@@ -777,18 +875,43 @@ def paginate_room_events(self, room_id, from_key, to_key=None,
...
+                seen.add(chunk_id)
+
+                event_ids = self._simple_select_onecol_txn(
+                    txn,
+                    table="chunk_backwards_extremities",
+                    keyvalues={"chunk_id": chunk_id},
+                    retcol="event_id"
+                )
+
+                extremities.extend(e for e in event_ids if e not in extremities)
```
Comment (richvdh) [inline synapse/storage/stream.py]: should we make `extremities` be a set or something?

File: synapse/storage/stream.py [hunk 9]
```diff
@@ -62,24 +63,25 @@
...
 _EventDictReturn = namedtuple("_EventDictReturn", (
-    "event_id", "topological_ordering", "stream_ordering",
+    "event_id", "chunk_id", "topological_ordering", "stream_ordering",
 ))
 
 
 def lower_bound(token, engine, inclusive=False):
     inclusive = "=" if inclusive else ""
-    if token.topological is None:
+    if token.chunk is None:
```
Comment (richvdh) [inline synapse/storage/stream.py]: do we not need to handle tokens that have a `topological` field but not a `chunk`?

File: synapse/storage/events.py [hunk 10]
```diff
@@ -1130,7 +1131,7 @@ def event_dict(event):
                 values={
                     "stream_ordering": event.internal_metadata.stream_ordering,
                     "chunk_id": chunk_id,
-                    "topological_ordering": event.depth,
+                    "topological_ordering": topo,
                     "depth": event.depth,
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: (Done)

File: synapse/storage/stream.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/storage/stream.py]: Could do, it doesn't really matter I don't think since we enforce the limit anyway and this way was just easier. Can change if you think its worth it

File: synapse/storage/stream.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/storage/stream.py]: Oh yes, this was a bug in that to get the token before you take one from `stream_ordering` rather than `topological_ordering`. I guess it matters less in this new world as `topological_ordering` is unique.

File: synapse/storage/stream.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/storage/stream.py]: Possibly, I think I need to document that its an ordered list.

File: synapse/storage/stream.py (hunk 6 above)
Comment (richvdh) [reply synapse/storage/stream.py]: fair. just thought it would save us pulling 100 events out of the chunk if the limit was 100 and we had already got 99 from the previous chunk.
